File: solutions/lua/circular-buffer/1/circular-buffer.lua
return {
  new = function(_ --[[self]], capacity)
    return {
      buffer = setmetatable({}, { __index = table }),
      read = function(self)
        assert(#self.buffer > 0, "buffer is empty")
        return self.buffer:remove(1)
      end,
      write = function(self, val)
        assert(#self.buffer < capacity, "buffer is full")
        self.buffer:insert(val)
      end,
      forceWrite = function(self, val)
        if #self.buffer == capacity then self:read() end
        self:write(val)
      end,
      clear = function(self)
        self.buffer = setmetatable({}, { __index = table })
      end
    }
  end,
}

File: solutions/ruby/circular-buffer/1/circular_buffer.rb
# frozen_string_literal: true

# class CircularBuffer
class CircularBuffer
  class BufferEmptyException < StandardError; end
  class BufferFullException < StandardError; end

  def initialize(capacity) = (@capacity = capacity
                              clear)

  def clear = @buffer = []
  def write(val) = @buffer.size == @capacity ? raise(BufferFullException) : @buffer << val
  def write!(val) = @buffer.size == @capacity ? (read && write(val)) : write(val)
  def read = @buffer.empty? ? raise(BufferEmptyException) : @buffer.shift
end
